### PR TITLE
Set spirv.Source according to the opencl.ocl.version and opencl.cxx.version.

### PR DESCRIPTION
**Pull request by haonanya**
Align with https://github.com/llvm/llvm-project/commit/5fd5b8ada70d9cbdaa8cc5bea50a6314e3140364.

### CODE REVIEW

**Comment by haonanya** [comment]
@svenvh , @MrSidims, can you please take a look? Thanks very much.

File: lib/SPIRV/PreprocessMetadata.cpp [hunk 1]
```diff
@@ -294,12 +294,16 @@ void PreprocessMetadataBase::preprocessOCLMetadata(Module *M, SPIRVMDBuilder *B,
   // !{x} = !{i32 3, i32 102000}
   B->addNamedMD(kSPIRVMD::Source)
       .addOp()
-      .add(CLVer == kOCLVer::CL21 ? spv::SourceLanguageOpenCL_CPP
-                                  : spv::SourceLanguageOpenCL_C)
+      .add(M->getNamedMetadata(kSPIR2MD::OCLCXXVer) &&
+                   (CLVer == kOCLVer::CLCXX10 || CLVer == kOCLVer::CLCXX2021)
+               ? spv::SourceLanguageOpenCL_CPP
```
Comment (svenvh) [inline lib/SPIRV/PreprocessMetadata.cpp:299]: Perhaps we should consider using `SourceLanguageCPP_for_OpenCL` instead, but I'm not entirely sure.  `SourceLanguageOpenCL_CPP` is for the "old" OpenCL C++ that was part of OpenCL 2.2.

File: lib/SPIRV/OCLUtil.cpp [hunk 2]
```diff
@@ -788,26 +788,39 @@ decodeOCLVer(unsigned Ver) {
...
+  unsigned OCLCXXVer = NamedMDOCLCXXVer ? GetVer(NamedMDOCLCXXVer) : 0;
+  // Check if OCLCXXVer is compatible with OCLVer
+  if (OCLVer && OCLCXXVer) {
+    if ((OCLVer == kOCLVer::CL20 && OCLCXXVer == kOCLVer::CLCXX10) ||
+        (OCLVer == kOCLVer::CL30 && OCLCXXVer == kOCLVer::CLCXX2021))
+      return OCLCXXVer;
+    report_fatal_error(llvm::Twine(
+        "opencl cxx version is not compatible with opencl c version!"));
+  }
+  return OCLVer ? OCLVer : OCLCXXVer;
```
Comment (svenvh) [inline lib/SPIRV/OCLUtil.cpp:823]: Will we ever have an OCLCXXVer without an OCLVer?  If not, we can just do

```suggestion
  return OCLVer;
```

File: lib/SPIRV/PreprocessMetadata.cpp (hunk 1 above)
Comment (haonanya) [reply lib/SPIRV/PreprocessMetadata.cpp:299]: +@MrSidims, @asudarsa, there are several places using SourceLanguageOpenCL_CPP, should we change to SourceLanguageCPP_for_OpenCL? Thanks very much.

Comment (MrSidims) [reply lib/SPIRV/PreprocessMetadata.cpp:299]: I'm also slightly leaning towards SourceLanguageCPP_for_OpenCL, technically it doesn't bring any semantic changes and from debugger perspective CPP_for_OpenCL and OpenCL_CPP are semantically equivalent.

**Review by MrSidims — APPROVED** [review]

File: lib/SPIRV/PreprocessMetadata.cpp (hunk 1 above)
Comment (svenvh) [reply lib/SPIRV/PreprocessMetadata.cpp]: We can do this as a separate PR, so not blocking this PR on that.

**Review by svenvh — APPROVED** [review]

**Comment by haonanya** [comment]
@MrSidims , @svenvh, can we land the PR? Thanks.

File: lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -4830,7 +4831,7 @@ void SPIRVToLLVM::transSourceLanguage() {
   SPIRVWord Ver = 0;
   SourceLanguage Lang = BM->getSourceLanguage(&Ver);
   if (Lang != SourceLanguageUnknown && // Allow unknown for debug info test
-      Lang != SourceLanguageOpenCL_C && Lang != SourceLanguageOpenCL_CPP)
+      Lang != SourceLanguageOpenCL_C && Lang != SourceLanguageCPP_for_OpenCL)
```
Comment (haonanya) [inline lib/SPIRV/SPIRVReader.cpp:4834]: Hi, @svenvh , @MrSidims , can we preserve old SourceLanguageOpenCL_CPP metadata? Currently spirv reader drop SourceLanguageOpenCL_CPP for sycl source https://github.com/intel/llvm/blob/sycl/clang/lib/CodeGen/CodeGenModule.cpp#L1360,L1362 on spirv->LLVM IR.

**Pull request merged**